hydromt_sfincs/sfincs.py

```diff
@@ -1011,6 +1016,126 @@ def setup_cn_infiltration(self, cn, antecedent_moisture="avg", reproj_method="me
         self.config.pop("qinf", None)
         self.set_config(f"{mname}file", f"sfincs.{mname}")
 
+    # Function to create curve number for SFINCS including recovery term (Kr)
+    def setup_curvenumber_infiltration_withrecovery(
+        self, data_catalog, name_files, reclass_table, effective, block_size=2000
+    ):
+        """Setup model the Soil Conservation Service (SCS) Curve Number (CN) files for SFINCS
+        including recovery term based on the soil saturation
+
+        Parameters
+        ---------
+        data_catalog    : data catalog used in the analysis
+        name_files      : list with name of the landcover, hydrologic soil group (HSG) and Saturated Hydraulic Conductivity (Ksat)
+        reclass_table   : reclass table to relate landcover with soiltype
+        effective       : float, estimate of percentage effective soil, e.g. 0.50 for 50%
+        block_size      : float, maximum block size - use larger values will get more data in memory but can be faster, default=500
+        """
+
+        # Read the datafiles
+        data_catalog = hydromt.DataCatalog(data_libs=data_catalog)
+        da_landuse = data_catalog.get_rasterdataset(
+            name_files[0], geom=self.region, buffer=10
+        )  # landcover
+        da_HSG = data_catalog.get_rasterdataset(
+            name_files[1], geom=self.region, buffer=10
+        )  # HSG
+        da_Ksat = data_catalog.get_rasterdataset(
+            name_files[2], geom=self.region, buffer=10
+        )  # Ksat
+        df_map = self.data_catalog.get_dataframe(reclass_table, index_col=0)
+
+        # Define outputs
+        da_smax = xr.full_like(self.mask, -9999, dtype=np.float32)
+        da_kr = xr.full_like(self.mask, -9999, dtype=np.float32)
+
+        # Define the blocks
+        nrmax = block_size
+        nmax = np.shape(self.mask)[0]
+        mmax = np.shape(self.mask)[1]
+        grid_dim = (nmax, mmax)
+        dx = self.config["dx"]
+        refi = dx / 30  # finest resolution of NLCD is ~30m
+        nrcb = int(np.floor(nrmax / refi))  # nr of regular cells in a block
+        nrbn = int(np.ceil(nmax / nrcb))  # nr of blocks in n direction
+        nrbm = int(np.ceil(mmax / nrcb))  # nr of blocks in m direction
+        x_dim, y_dim = self.mask.raster.x_dim, self.mask.raster.y_dim
```

Comment:
Do we nee to compute the blocks here based on the landuse data resolution or can we simply define `block_size` based on the SFINCS grid here? If we need the first, it would be good to base `refi` on the actual landuse resolution. Otherwise we can get rid of this block I think and work with `nrcb = block_size`

Reply:
We do need to compute it based on the land cover since that is what is going to determine how much memory we are going to use. I did make it variable based on the landcover dataset.